### PR TITLE
'XAML Suggested Action' extensibility

### PR DESCRIPTION
**Pull request by koljaka**
Add documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ more recent surface isolation architecture. [This article](./documents/xaml-desi
 
 - [XAML designer extensibility migration](./documents/xaml-designer-extensibility-migration.md)
 - [Toolbox population](./documents/toolbox-population.md)
+- [XAML Suggested Actions Extensibility](./documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md)
```
Comment (debonte) [inline README.md:10]: The word XAML seems unnecessary here since we're in the "XAML Designer Extensibility" repo.
Also "Extensibility" should probably be lowercase. I can understand Suggested Actions being uppercase since it's a feature name.

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
+## Suggested Actions Extensibility
+
+- [Tokens](./xaml-designer-suggested-actions-extensibility-tokens.md)
+- [Features](./xaml-designer-suggested-actions-extensibility-features.md)
+- [Behaviors](./xaml-designer-suggested-actions-extensibility-behaviors.md)
+- [Customization](./xaml-designer-suggested-actions-extensibility-customization.md)
+
+Starting from Visual Studio 2019 16.7 Preview 3 we added extensibility support for "XAML Suggested Actions"
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:8]: "Starting from" -> "In"

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+>*Note: this API will be available only in Preview Channel until feature is finalized and completed*
+
+To enable "Xaml Suggested Actions" for any control, `SuggestedActionProvider` feature provider should be created and registered in metadata.
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:11]: "`SuggestedActionProvider` feature provider should be created" -> "a `SuggestedActionProvider` feature provider should be created"

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+>*Control could have multiple Suggested Actions providers. Each separate provider will be shown as a separate tab in UI.*
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:13]: Control -> A control
Each separate -> Each
in UI -> "in the UI" or maybe something more specific like "in the Suggested Actions UI"?

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-behaviors.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+```cs
+public class ExampleBehaviorActionProvider : BehaviorActionProvider
+{
+    //Specify Behavior attached property name
+    public override string BehaviorPropertyName => "Triggers";
+
+    //Specify fully qualified type name of a class that implements a Behavior attached property
+    public override string BehaviorPropertyTypeFullName => "Microsoft.Xaml.Behaviors.Interaction";
+
+    public override string Header => "Microsotf Xaml Behaviors";
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-behaviors.md:16]: Microsotf -> Microsoft
Also fix the image that shows the misspelling in xaml-designer-suggested-actions-extensibility.md

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    public static ActionToken Token_Group_Common        = SuggestedActionProviderTokens.Token_Group_Common;
+    public static ActionToken Token_Group_Specific      = SuggestedActionProviderTokens.Token_Group_Specific;
+    public static ActionToken Token_Property_Background = SuggestedActionProviderTokens.Token_Property_Background;
+    public static ActionToken Token_Property_Content    = SuggestedActionProviderTokens.Token_Property_Content;
+    public static ActionToken Token_Property_IsEnabled  = SuggestedActionProviderTokens.Token_Property_IsEnabled;
+    public static ActionToken Token_Property_IsCancel   = new ActionToken(0x1001);
+    public static ActionToken Token_Property_IsDefault  = new ActionToken(0x1002);
+    public static ActionToken Token_Last = new ActionToken(0x10FF);
+    
+    public override string Header { get => "Actions"; }
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:32]: Can this property syntax
public override string Header { get => "Actions"; }
be simplified to 
public override string Header => "Actions";
?

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        this.AddGroup(new ActionGroup(ButtonActionProvider.Token_Group_Specific,
+            new PropertyAction(ButtonActionProvider.Token_Property_IsCancel, "IsCancel"),
+            new PropertyAction(ButtonActionProvider.Token_Property_IsDefault, "IsDefault")));
+        this.AddGroup(new BorderBrushActionGroup());
+        this.AddGroup(new VisibilityActionGroup());
+        this.AddGroup(new FontSettingsActionGroup());
+    }
+}
+```
+### Example for `MyButtonSuggestedActionProvider : ButtonActionProvider`
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:53]: I'm not sure what this means. It seems unrelated to the image above? Seems like this needs more explanation.

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+builder.AddCustomAttributes("System.Windows.Controls.Button", new FeatureAttribute(typeof(ButtonActionProvider)));
+
+//Same for another control
+builder.AddCustomAttributes("CustomControlLibrary.MyButton", new SuggestionsAttribute());
+builder.AddCustomAttributes("CustomControlLibrary.MyButton", new FeatureAttribute(typeof(MyButtonSuggestedActionProvider)));
+...
+```
+
+### Documentation link
+At the top of 'XAML Suggested Actions' there is a Type Name label.
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:104]: "At the top of 'XAML Suggested Actions'" should be something like "At the top of the Suggested Actions dialog in the designer". Maybe including an image here would help clarify.

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-tokens.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
+## Suggested Actions Tokens
+
+Each action group and action could have assigned token of type `ActionToken`. Using this token, action group or action could be referenced and modified. Example:
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-tokens.md:3]: "could have assigned token" -> "could have an assigned token"
" Using this token, action group or action could" ->  "Using this token, the corresponding action group or action can"

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+Example for Button:
+
+![extensibility-migration-architecture](xaml-suggested-actions.png)
+
+### Implementation for example above:
+```CS
+public class ButtonActionProvider : SuggestedActionProvider
+{
+    public static ActionToken Token_Group_Common        = SuggestedActionProviderTokens.Token_Group_Common;
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:23]: Why are you creating aliases for the well-known tokens? Seems unnecessarily verbose.

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
+## Suggested Actions Features
+
+Currently we have 5 types of actions that could be added:
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:3]: "we have" -> "there are"
Maybe replace "Currently" with "In 16.7"?

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+{
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+    
+    public string MyCustomActionText { get; set; }
+    public ICommand ResetCommand { get; set; }
+    
+    public MyCustomAction()
+    {
+        this.Template = res["MyCutsomActionTemplate"] as ControlTemplate;
+        this.MyCustomActionText = "Change Content oprions:";
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:64]: oprions -> options

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+```
+![extensibility-migration-architecture](xaml-suggested-actions-selectableAction.png)
+
+## *CustomAction* - action with custom template
+```cs
+public class MyCustomAction : CustomAction
+{
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+    
+    public string MyCustomActionText { get; set; }
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:58]: Should these properties be readonly? { get; }

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    private class ResetCommandImpl : ICommand
+    {
+        public event EventHandler CanExecuteChanged;
+        public bool CanExecute(object parameter)
+        {
+            return true;
+        }
+        public void Execute(object parameter)
+        {
+            (parameter as ModelItem).Content.SetValue("Content");
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:77]: Switch to hard cast or add null check

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+}
+```
+XAML from `SuggestedActionsResources` ResourceDictionary:
+```xml
+<ControlTemplate x:Key="MyCutsomActionTemplate">
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:84]: Cutsom -> Custom

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 16]
```diff
@@ -0,0 +1,227 @@
+## Suggested Actions Customization
+
+There are multiple ways you could customize the view of "XAML Suggested Actions":
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:3]: "you could" -> "that you can"

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+### 1. Action
+![extensibility-migration-architecture](xaml-suggested-customAction.png)
+
+To make custom action use `CustomAction` class:
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:8]: "To make custom action use" -> "To make a custom action use the"

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+public class MyCustomAction : CustomAction
+{
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+    
+    public string MyCustomActionText { get; set; }
+    public ICommand ResetCommand { get; set; }
+    
+    public MyCustomAction()
+    {
+        this.Template = res["MyCutsomActionTemplate"] as ControlTemplate;
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:20]: Cutsom -> Custom

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+    private class ResetCommandImpl : ICommand
+    {
+        public event EventHandler CanExecuteChanged;
+        public bool CanExecute(object parameter)
+        {
+            return true;
+        }
+        public void Execute(object parameter)
+        {
+            (parameter as ModelItem).Content.SetValue("Content");
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:34]: Hard cast or add null check

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+        }
+    }
+}
+```
+```xml
+<ControlTemplate x:Key="MyCutsomActionTemplate">
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:40]: Cutsom -> Custom

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+
+    public PropertyAction BackgroundProperty { get; set; }
+    public LinkAction HelpLinkAction { get; set; }
+    public CommandAction ShowDialogCommandAction { get; set; }
+    public SelectableAction SetBackgroundToAction { get; set; }
+
+    public MyCustomActionGroup()
+    {
+        this.Template = res["MyCutsomActionGroupTemplate"] as ControlTemplate;
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:77]: Cutsom -> Custom

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+            PropertyIdentifier backgrounPropertyIdentifier = new PropertyIdentifie(new TypeIdentifier("System.Windows.Controls.Control"), "Background");
+            this.ModelItem.Properties[backgrounPropertyIdentifier].Collection.Ad(solidColorBrush);
+        };
+        
+        this.SetBackgroundToAction.SetActionProvider(this.ActionProvider);
+    }
+}
+```
+```xml
+<ControlTemplate x:Key="MyCutsomActionGroupTemplate">
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:120]: Cutsom -> Custom

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+### 3. SuggestedActionProvider
+![extensibility-migration-architecture](xaml-suggested-customSuggestedActionProvider.png)
+```cs
+public class MyCustomSuggestedActionProvider : SuggestedActionProvider
+{
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+
+    public override string Header => "Custom SuggestedActionProvider";
+    public override string Type => "System.Windows.Controls.Button";
+    public string ShortDocumentation { get; set; }
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:157]: Should ShortDocumentation property by readonly { get; }?

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+### 4. ActionProviderBase 
+![extensibility-migration-architecture](xaml-suggested-customActionProviderBase.png)
+```cs
+public class MyCustomActionProvider : ActionProviderBase
+{
+    private SuggestedActionsResources res = new SuggestedActionsResources();
+
+    public override string Header => "Documentation";
+    public override string Type => "System.Windows.Controls.Grid";
+    public string Documentation { get; set; }
```
Comment (debonte) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-customization.md:202]: Should Documentation property be readonly { get; }?

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md (hunk 14 above)
Comment (debonte) [reply documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility-features.md:77]: Looks like this one didn't get fixed.

File: documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+            new PropertyAction(ExampleButtonSuggestedActionProvider.Token_Property_IsCancel,"IsCancel"),
+            new PropertyAction(ExampleButtonSuggestedActionProvider.Token_Property_IsDefault,"IsDefault")));
+        
+        this.AddGroup(new BorderBrushActionGroup());
+        this.AddGroup(new VisibilityActionGroup());
+        this.AddGroup(new FontSettingsActionGroup());
+    }
+}
+```
+### Example for inherited controls
```
Comment (mgoertz-msft) [inline documents/xaml-suggested-actions/xaml-designer-suggested-actions-extensibility.md:52]: Perhaps you could add a sentence here explaining how the provider can modify the base behavior. I had to look back and forth a few time to notice the differences in the image.
